File: 2024/01-solution.R
##################################################
## load required libraries #######################
##################################################

library(readr) ## for reading in csvs
library(dplyr) ## for data manipulation

##################################################
## read in data ##################################
##################################################

data <- read_delim("01-input.txt",
                   delim = "   ",
                   col_names = c("A", "B"),
                   show_col_types = FALSE)

##################################################
## process data ##################################
##################################################

## pull out sorted first column
A <- sort(data$A, decreasing = FALSE)

## pull out sorted second column
B <- sort(data$B, decreasing = FALSE)

## smoosh them back together
updated_data <- cbind.data.frame(A, B)

## calculate distance
updated_data$distance <- abs(updated_data$A - updated_data$B)

##################################################
## solution: part A ##############################
##################################################

sum(updated_data$distance)

##################################################
## solution: part B ##############################
##################################################

## count the number of times each element of B occurs, just look at 
## items that also occur in A
count_table <- table(B)[as.character(A)]

## calculate the weird similarity score
sum(as.numeric(count_table)*as.numeric(names(count_table)), na.rm = TRUE)

File: 2024/01-solution.py
##################################################
## load required packages ########################
##################################################

import pandas as pd
import numpy as np

##################################################
## read in data ##################################
##################################################

data = pd.read_csv('01-input.txt',
                   sep = "   ",
                   header = None,
                   engine = 'python')

##################################################
## process data ##################################
##################################################

## pull out sorted first column
A = data[0].sort_values().reset_index(drop = True)

## pull out sorted second column
B = data[1].sort_values().reset_index(drop = True)

## smoosh them back together
updated_data = pd.concat([A, B], axis = 1)
updated_data.columns = ["A", "B"]

## calculate distance
updated_data["distance"] = abs(updated_data["A"] - updated_data["B"])

##################################################
## solution: part A ##############################
##################################################

#print(sum(updated_data["distance"]))

##################################################
## solution: part B ##############################
##################################################

## move over to numpy
a = A.values
b = B.values

## use the histogram function to calculate counts per item
a_counts = np.bincount(a)
b_counts = np.bincount(b)

print(sum(a_counts[:len(b_counts)]*b_counts*np.arange(len(b_counts))))